core/src/main/scala/org/apache/spark/UploadDirManager.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark
+
+import org.apache.hadoop.fs.{FileSystem, Path}
+
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.{ShutdownHookManager, Utils}
+
+/**
+ * Kubernetes upload directory manager. It's responsible for cleanup of the upload directory
+ * when spark job finishes.
+ */
+private[spark] class UploadDirManager(conf: SparkConf, appId: String) extends Logging {
```

Comment:
I wonder if we can share logic here with YARN? Since I imagine Spark on YARN also needs to cleanup these jars.

Reply:
The staging directory provided by spark.kubernetes.file.upload.path is specific to K8s. Though Yarn has already got logic to cleanup it's staging directory. That is present in ApplicationMaster.scala.

Reply:
Why uploadDirManager not placed in KubernetesClientApplication? When driver pod status is failed, you can delete k8s upload directory.

Reply:
It is being registered in KubernetesClientApplication also.

Reply:
I mean, this cleanup is only for the kubernetes.It will be clearer to put all in the k8s module.You can tell by the status of driver pod if a cleanup is needed.

Reply:
The reason it's not put in k8s module is because cleanup can happen in two cases:
1. Spark app submission fails and clean up is required -> THis is handled through KubernetesClientApplication.
2. Spark app submission succeeds and driver is launched in client mode. Clean up is required when job finishes -> This is handled in SparkSumbit, which is not in k8s module. This scenario can't be handled in  KubernetesClientApplication as it's not called on the driver side.